### PR TITLE
Provide script for packaging mlcubes

### LINKED ISSUE #407: Documentation on how to deploy MLCubes

### LINKED ISSUE #410: Provide Script to package and prepare an MLCube for medperf deployment

### PR DESCRIPTION
**Pull request by aristizabal95**
This PR creates a simple script for automatically preparing and packaging mlcubes for deployment. It by default creates a `./deploy` path inside the `mlcube` folder, with all the files that need to be hosted separately for medperf. Additionally, passing the `--output` option creates a tarball file with the contents of the `deploy` folder. Closes #410. Closes #407.

### CODE REVIEW

**Comment by aristizabal95** [comment]
Add validation steps to the script:
- parse the `mlcube.yaml` file
- check runner configurations
  - give suggestions (e.g remember to push image to registry
  - check for ambiguity (docker image and singularity image specified). throw error if that happens
- search for valid configurations
- if any configuration is valid, validate the parameters of that configuration
- if no configuration is valid, throw an error
- let the user know what valid configurations were found
- check for specified assets and throw an error if they're not there
- If everything goes well, package the contents as done right now (don't forget to package singularity image if present)

**Comment by hasan7n** [comment]
@aristizabal95 it is looking great! fantastic!

here are the things I found:

1. Workspace folder: it is not always needed (people can just submit an mlcube.yaml file), but the script raises if it doesn't exist. Actually I think when the script checks for the other files (like params file, singularity image...) it will check the existence of the workspace folder in the process, so no need to check it beforehand.
2. results file and statistics file: there is no need to force a specific name. medperf overrides these parameters when running.
3. I don't see validation of additional files. so the script will not raise error if it didn't find the additional files folder, even if the mlcube config needs it.

4. Do you think using the word "deploy" is suitable? It gives the impression that the files will be automatically hosted. what about just "prepare" or "package"? without using "deploy"
5. The script doesn't print to the user what medperf tasks it detected. (or, something like: "hey your mlcube can be used as data prep and evaluator" )


**note:** MLCube doesn't seem to allow specifying singularity config without an image name. It throws an error. This has to be communicated to the best practices team, but we can keep our logic here as is.

**Comment by aristizabal95** [comment]
Hey @hasan7n  thanks for the suggestions! I just made all the requested changes. Let me know if it looks okay!

File: scripts/package-mlcube.py [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+from utils.package import package
+from utils.validate import is_valid
+
+
+def get_args():
+    desc = (
+        "Helper script to prepare an MLCube for medperf/challenge submission.\n"
+        + "This script will create a 'deploy' path with the files that need to be hosted\n"
+        + "for medperf mlcube submission. Additionally, if an output path is provided,\n"
+        + "all deployment files are packaged into the output tarball file. Useful for competitions"
```
Comment (hasan7n) [inline scripts/package-mlcube.py:11]:
```suggestion
        + "all deployment files are packaged into the output tarball file."
```

**Comment by hasan7n** [comment]
@aristizabal95 Packaging needs validation, and validation needs parsing the manifest file. So it's difficult to write the code somewhere in between "no validation" and "full validation", so I chose full validation.
Regarding our previous concern of code maintainability/readability, I found this solution https://pypi.org/project/schema/. I did the necessary changes, please check and let me know what do you think about the code clarity/readability.

Also finalized the documentation. If you are ok with everything, we can merge

File: docs/concepts/mlcube_files.md [hunk 2]
```diff
@@ -1 +1,61 @@
...
+python scripts/package-mlcube.py \
+   --mlcube path/to/mlcube \
+   --mlcube-types <list-of-comma-seprated-strings> \
+   --output path/to/file.tar.gz
+```
+
+where:
+
+  - `path/to/mlcube` is the path to the MLCube folder containing the manifest file (`mlcube.yaml`)
+  - `--mlcube-types` specifies a comma-separated list of MLCube types (e.g. is it supposed to be a Preparation MLCube, Model MLcube, ...)
```
Comment (aristizabal95) [inline docs/concepts/mlcube_files.md:49]: I think we need to show examples of the strings that can be passed to this parameter

File: docs/gen_ref_pages.py [hunk 3]
```diff
@@ -28,9 +30,7 @@
             continue
         if parts == ():
             continue
-        if parts[0] == "tests":
-            continue
-        if parts[0] == "templates":
+        if parts[0] in exclude_paths:
```
Comment (aristizabal95) [inline docs/gen_ref_pages.py:33]: Thanks for this haha

File: scripts/utils/schemas.py [hunk 4]
```diff
@@ -0,0 +1,144 @@
...
+    def validate(self, data, **kwargs):
+        files = kwargs.pop("files")
+        files.add(ADD_PATH)
+        return super().validate(data, **kwargs)
+
+
+# Setup extra inputs validators: params file and additional files
+
+params_validator = And(
+    Use(parse_value),
```
Comment (aristizabal95) [inline scripts/utils/schemas.py:36]: Can you for how these small validators work or what they're accomplishing? Just to make it easy to read by someone that doesn't know how schema works (in this case me haha)

File: scripts/utils/schemas.py [hunk 5]
```diff
@@ -0,0 +1,144 @@
...
+            **wild_card,
+        },
+        **wild_card,
+    },
+    name="Metrics MLCube",
+)
+
+# Put the available schemas in a dict
+schemas = {
+    "prep": PREP_MLCUBE_TASKS,
```
Comment (aristizabal95) [inline scripts/utils/schemas.py:140]: If these are the mlcube types that are exposed to the user, I'd probably use the complete names for prep `data-preparator` and `data-preparator-separate-labels`. Might be a mouthful, but I think it's best to be clear and explicit here

**Review by aristizabal95 — COMMENTED** [review]
Hey Hasan! This is looking great! I really like how tasks are being validated now using the `schema` library, it looks clean and easy to comprehend! I just left a few comments on a few UX details. I can't approve because I'm the author of the PR

File: docs/gen_ref_pages.py (hunk 3 above)
Comment (hasan7n) [reply docs/gen_ref_pages.py:33]: thansk to you ! :) (you did that change haha)

**Review by hasan7n — APPROVED** [review]

**Comment by hasan7n** [comment]
@aristizabal95 Thanks! I did the requested changes. Please merge if you are okay with it

**Review by aristizabal95 — COMMENTED** [review]
Did you make any changes regarding the comments I left? I'm asking because I couldn't see noticeable changes. In any case, my comments were more recommendations, and I think everything looks great at this stage. We can merge this!

**Pull request merged**